### PR TITLE
[PREVIEW] detect/lua: Cleanup thread contextual info when needed

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #4740

I used these rules -- the first rule demonstrated the issue. With this PR, memory is no longer overrun.
```
alert http any any -> any any (msg:"High entropy in file name download"; file.name; content:".exe"; endswith; fast_pattern;  bsize:<21; lua:any.lua; xbits: isset, urlentropy.detected; track ip_dst, expire 3600; target: dest_ip; sid:10; rev:1;)
alert http any any -> any any (msg:"High entropy in file name download"; file.name; content:".exe"; endswith; fast_pattern;  bsize:<21; lua:any.lua;   target: dest_ip; sid:11; rev:1;)
```

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [3273](https://redmine.openinfosecfoundation.org/issues/3273)

Describe changes:
- Modified `Free` func to require an additional param: `DetectEngineCtx *`. This had a huge ripple.

### CODE REVIEW

File: src/app-layer-detect-proto.c [hunk 1]
```diff
@@ -1619,7 +1619,7 @@ void AppLayerProtoDetectPPRegister(uint8_t ipproto,
     SCEnter();
 
     DetectPort *head = NULL;
-    DetectPortParse(NULL,&head, portstr);
+    DetectPortParse(NULL, &head, portstr);
```
Comment (victorjulien) [inline src/app-layer-detect-proto.c:1622]: unrelated style fix

File: src/app-layer-detect-proto.c [hunk 2]
```diff
@@ -1395,13 +1395,13 @@ static int AppLayerProtoDetectPMPrepareMpm(AppLayerProtoDetectPMCtx *ctx)
     SCReturnInt(ret);
 }
 
-static void AppLayerProtoDetectPMFreeSignature(AppLayerProtoDetectPMSignature *sig)
+static void AppLayerProtoDetectPMFreeSignature(DetectEngineCtx *de_ctx, AppLayerProtoDetectPMSignature *sig)
```
Comment (victorjulien) [inline src/app-layer-detect-proto.c:1398]: we have no detect engine in the proto detection, so no need to pass it around. Just pass NULL into DetectContentFree

File: src/suricata.c [hunk 3]
```diff
@@ -336,6 +336,7 @@ static void GlobalsDestroy(SCInstance *suri)
     }
     DetectEnginePruneFreeList();
 
+    // XXX Prefer to have de_ctx available here
```
Comment (victorjulien) [inline src/suricata.c:339]: Not needed AFAICS. The app-layer doesn't use a detect engine, so pass NULL to any detect keyword free func that might it

**Comment by victorjulien** [comment]
Btw the typedef in detect-tag.h can be removed, instead you can add `#include "detect.h"`

**Comment by jlucovsky** [comment]
Continued in #4843
